Enable Walkthroughs

dhruvisompura opened the pull request:
Addresses #8054 #3406

This PR enables walkthroughs for Positron again and modifies which walkthroughs can be shown on the welcome page. All walkthroughs should be accessible via the command palette action `welcome.showAllWalkthroughs` (aka "Welcome: Open Walkthrough"). 

We can make changes to `rankWalkthrough` so walkthroughs we want to show on the welcome page have a rank and all other walkthroughs are not given a rank.  We will rely on the `source` and `id` fields to identify walkthroughs.

Note: https://github.com/posit-dev/positron/pull/7998 added the change which ensures a "featured walkthrough" is not automatically launched when a user opens Positron for the first time.

#### Technical Details
Walkthroughs are implemented through the following:
- **WalkthroughsService**: Manages walkthrough data, walkthrough progress, and events
- **GettingStartedPage**: The UI component that renders walkthroughs (and welcome page)
- **gettingStartedContent.ts**: The file where built-in walkthroughs are defined

walkthroughs which have been registered and are ready to be rendered have the following definition:
```typescript
export interface IWalkthrough {
	id: string; // unique identifier for the walkthrough
	title: string; // display name/walkthrough heading shown in UI
	description: string; // summary of walkthrough that is displayed under the title in the welcome page
	order: number; // determines position in walkthrough list (Used by `rankWalkthrough` to calculate final ordering)
	source: string; // identifies where walkthrough is from (values can be "Built-In" or the extension display name)
	isFeatured: boolean; // If true, the walkthrough is given visual prominence in the UI, appear higher up list (receive a bonus in `rankWalkthrough`)
	next?: string; // ID of the walkthrough that should be suggested after completing this one. Creates a guided path through multiple walkthroughs
	when: ContextKeyExpression; // A context condition that determines whether the walkthrough should be shown
	steps: IWalkthroughStep[]; // content of walkthrough
	icon:
	| { type: 'icon'; icon: ThemeIcon }
	| { type: 'image'; path: string };
	walkthroughPageTitle: string; // for web, title displayed at the top of the browser tab when walkthrough is open
}
```

**Command Flow**

The command to open a specific walkthrough is `workbench.action.openWalkthrough` in `gettingStarted.contribution.ts`:

1. User triggers the command with a walkthrough ID (and optionally a step ID)
2. The command handler finds or creates a `GettingStartedInput` editor (used by `GettingStartedPage` component)
3. It opens this input in an editor pane, showing the walkthrough UI

**UI Rendering**

When a walkthrough is opened:

1. `GettingStartedPage.buildCategorySlide()` builds the UI for the walkthrough
2. The steps are rendered with media, descriptions, and navigation controls (this adds the handlers to track the walkthrough progress as well which was previously broken)
3. The user can navigate through steps using UI controls or keyboard

**Step Progression**

Steps can be completed in several ways:

1. Manually by the user clicking on a completion checkbox
2. Automatically through completion events defined in the step configuration
3. Programmatically via commands like `welcome.markStepComplete`

The `WalkthroughsService` maintains the state of walkthrough steps and tracks which steps are completed. The `GettingStartedPage` component maintains the UI state of the current walkthrough and updates the UI when a step changes. 

Note: The `buildGettingStartedWalkthroughsList` function in `GettingStartedPage` is responsible for creating and populating the list of available walkthrough dom elements. Without this list, UI operations like focusing, and traversing the walkthroughs don't work as expected.

**Welcome Page Integration**

Walkthroughs appear on the welcome page when the welcome page editor is opened:

1. The welcome page is an instance of the `GettingStartedPage`
2. Walkthroughs are displayed in a list when `buildGettingStartedWalkthroughsList()` is called

### Release Notes

#### New Features

- Re-enable walkthroughs in Positron (#3406)
- Only display Positron featured walkthroughs on welcome page (#8054)

#### Bug Fixes

- N/A


### Screenshots


https://github.com/user-attachments/assets/073a9d05-cc76-45e8-9565-f51bdf1649d7



### QA Notes

@:web @:win @:welcome

dhruvisompura committed:
Enable walkthroughs again

dhruvisompura committed:
Prevent featured walkthrough from launching on first startup

dhruvisompura committed:
Filter out all built in walkthroughs

juliasilge commented:
We have https://github.com/posit-dev/positron/pull/8227 merged in now as well, so the VS Code migration walkthrough could also be pulled into this PR to use on the Welcome page.

dhruvisompura committed:
Add walkthrough specific badge colors

Introduces a new theme color for walkthrough badges on the welcome page.

dhruvisompura committed:
Only show migration walkthrough guides on welcome page

dhruvisompura committed:
Add missing icon for positron-r extension

The icon will be featured next to walkthroughs on the welcome page

dhruvisompura committed:
Hide "Remote-SSH" action on welcome page for web

dhruvisompura committed:
Remove check for unused buildNewCategorySlide fn

The upstream method "buildNewCategorySlide" is used
and the check to avoid TS compiler errors can be removed

dhruvisompura commented on extensions/positron-r/package.json:
```diff
@@ -17,6 +17,7 @@
       "description": "%r.capabilities.untrustedWorkspaces.description%"
     }
   },
+  "icon": "icon.png",
```
The listing for this walkthrough on the welcome page was missing an icon. I added one to the extension so it doesn't show the "no image" icon.

dhruvisompura commented on src/vs/workbench/common/theme.ts:
```diff
@@ -471,6 +471,13 @@ export const ACTIVITY_BAR_BADGE_BACKGROUND = registerColor('activityBarBadge.bac
 	hcLight: '#0F4A85'
 }, localize('activityBarBadgeBackground', "Activity notification badge background color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."));
 
+export const WALKTHROUGH_ACTIVITY_BAR_BADGE_BACKGROUND = registerColor('walkthroughActivityBarBadge.background', {
+	dark: '#007ACC',
+	light: '#007ACC',
+	hcDark: '#000000',
+	hcLight: '#0F4A85'
+}, localize('walkthroughActivityBarBadge', "Activity notification badge background color for walkthroughs."));
+
```
The VS Code badge color is red but the mock ups we had used blue. Added this new theme color so we can customize the badge colors for walkthroughs

dhruvisompura commented on src/vs/workbench/contrib/welcomeGettingStarted/browser/gettingStarted.ts:
```diff
@@ -187,6 +189,7 @@ export class GettingStartedPage extends EditorPane {
 	private detailsRenderer: GettingStartedDetailsRenderer;
 
 	private readonly categoriesSlideDisposables: DisposableStore;
+	private showFeaturedWalkthrough = true;
```
these vars had been removed when we turned walkthroughs off initially. Now that we have re-introduced walkthroughs, we need them back

dhruvisompura commented on src/vs/workbench/contrib/welcomeGettingStarted/browser/gettingStarted.ts:
```diff
@@ -999,13 +1000,32 @@ export class GettingStartedPage extends EditorPane {
 			const telemetryNotice = $('p.telemetry-notice');
 			this.buildTelemetryFooter(telemetryNotice);
 			footer.appendChild(telemetryNotice);
+		} else if (!this.productService.openToWelcomeMainPage && !someStepsComplete && !this.hasScrolledToFirstCategory && this.showFeaturedWalkthrough) {
+			// Always show index page instead of automatically opening the first walkthrough category
+			// --- Start Positron ---
+			/*
+			const firstSessionDateString = this.storageService.get(firstSessionDateStorageKey, StorageScope.APPLICATION) || new Date().toUTCString();
+			const daysSinceFirstSession = ((+new Date()) - (+new Date(firstSessionDateString))) / 1000 / 60 / 60 / 24;
+			const fistContentBehaviour = daysSinceFirstSession < 1 ? 'openToFirstCategory' : 'index';
+
+			if (fistContentBehaviour === 'openToFirstCategory') {
+				const first = this.gettingStartedCategories.filter(c => !c.when || this.contextService.contextMatchesRules(c.when))[0];
+				if (first) {
+					this.hasScrolledToFirstCategory = true;
+					this.currentWalkthrough = first;
+					this.editorInput.selectedCategory = this.currentWalkthrough?.id;
+					this.buildCategorySlide(this.editorInput.selectedCategory, undefined);
+					this.setSlide('details');
+					return;
+				}
+			}
+			*/
+			// --- End Positron ---
```
We had removed this code completely last time we had removed walkthroughs but I am adding it back and commenting out the logic mainly so its easier to tweak the behavior in the future.

dhruvisompura commented on src/vs/workbench/contrib/welcomeGettingStarted/browser/gettingStarted.ts:
```diff
@@ -1239,6 +1254,25 @@ export class GettingStartedPage extends EditorPane {
+			 * and the IDs are hardcoded here and need to be kept in sync with the Id
+			 * of the walkthroughs in the Positron extensions.
+			 */
+			const allowedWalkthroughIds = ['ms-python.python#positron.migrateFromVSCode', 'positron.positron-r#positron.r.migrateFromRStudio'];
+
+			if (allowedWalkthroughIds.every(walkthroughId => walkthroughId !== e.id)) {
+				return null;
+			}
```
I don't love that these walkthrough IDs are hardcoded here and would prefer finding an alternative to this in the near future. I'll document my thoughts and findings in #3406. The main concern is the ID changing and us forgetting to update the reference here.

softwarenerd reviewed (dismissed):
Ran the walkthrough. Looks good from here.

dhruvisompura committed:
Merge branch 'main' into feature/enable-walkthroughs

dhruvisompura committed:
Enable e2e test check for remote-ssh command on welcome page
